Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -60,50 +58,39 @@ public function __construct()
 
     /**
      * @param string $identifier
-     * @return AggregateRootInterface
+     * @return EventSourcedAggregateRootInterface
      * @throws AggregateRootNotFoundException
      */
     final public function findByIdentifier($identifier)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
-        try {
-            $eventStream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            return null;
-        }
-
         if (!class_exists($this->aggregateClassName)) {
             throw new AggregateRootNotFoundException(sprintf("Could not reconstitute the aggregate root %s because its class '%s' does not exist.", $identifier, $this->aggregateClassName), 1474454928115);
         }
         if (!is_subclass_of($this->aggregateClassName, EventSourcedAggregateRootInterface::class)) {
             throw new AggregateRootNotFoundException(sprintf("Could not reconstitute the aggregate root '%s' with id '%s' because it does not implement the EventSourcedAggregateRootInterface.", $this->aggregateClassName, $identifier, $this->aggregateClassName), 1474464335530);
         }
+        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
+        $eventStream = $this->eventStore->get($streamName);
         return call_user_func($this->aggregateClassName . '::reconstituteFromEventStream', $identifier, $eventStream);
     }
 
     /**
      * @param EventSourcedAggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    final public function save(EventSourcedAggregateRootInterface $aggregate)
+    final public function save(EventSourcedAggregateRootInterface $aggregate, int $expectedVersion = null)
     {
         $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
-        }
-
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getReconstitutionVersion() + 1;
+        $uncommittedEventsWithMetadata = array_map(function(EventInterface $event) use(&$version) {
+            return new EventWithMetadata($event, new EventMetadata([EventMetadata::VERSION => $version ++]));
+        }, $uncommittedEvents);
```

Comment:
I'm not happy that we increase the version here.. The version should only ever be set by the Store (it might not even be a sequence without gaps).. But IMO that can be improved in a separate change